8330542: Template for Creating Strict JAXP Configuration File

JoeWang-Java opened the pull request:
Add two sample configuration files:

  jaxp-strict.properties: used to set strict configuration, stricter than jaxp.properties in previous versions such as JDK 22

>   jaxp-compat.properties: used to regain compatibility from any more restricted configuration than previous versions such as JDK 22

Updated on 5/16/2024

Design change:
The design is changed to include in the JDK two configuration files that are the default jaxp.properties and jaxp-strict.properties, instead of three, dropping jaxp-compat.properties.

Updated on 5/18/2024

Withdraw changes to jaxp.properties. The original idea was to match jaxp-strict.properties with regard to the properties. However, that change impact the configuration process, resulting in tests that verify the process to fail.

Updated on 5/23/2024

Provide a template `jaxp-strict.template` instead of a properties file. This template can be used to create custom configuration files.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8331016](https://bugs.openjdk.org/browse/JDK-8331016) to be approved

### Issues
 * [JDK-8330542](https://bugs.openjdk.org/browse/JDK-8330542): Template for Creating Strict JAXP Configuration File (**Enhancement** - P4)
 * [JDK-8331016](https://bugs.openjdk.org/browse/JDK-8331016): Template for Creating Strict JAXP Configuration File (**CSR**)


### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**) ⚠️ Review applies to [714095d1](https://git.openjdk.org/jdk/pull/18831/files/714095d1fa542d8d012d500ced0032531aa9e3ef)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [2ee2c7ca](https://git.openjdk.org/jdk/pull/18831/files/2ee2c7ca54ccc79c3401aa407fe416ad15699d49)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [0de8ad69](https://git.openjdk.org/jdk/pull/18831/files/0de8ad6913565b91efb831176d21a5b100011032)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to [714095d1](https://git.openjdk.org/jdk/pull/18831/files/714095d1fa542d8d012d500ced0032531aa9e3ef)
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [714095d1](https://git.openjdk.org/jdk/pull/18831/files/714095d1fa542d8d012d500ced0032531aa9e3ef)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18831/head:pull/18831` \
`$ git checkout pull/18831`

Update a local copy of the PR: \
`$ git checkout pull/18831` \
`$ git pull https://git.openjdk.org/jdk.git pull/18831/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18831`

View PR using the GUI difftool: \
`$ git pr show -t 18831`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18831.diff">https://git.openjdk.org/jdk/pull/18831.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18831#issuecomment-2062722515)

AlanBateman commented:
Just some context here. The JDK has been on a path for many releases to make XML processing more secure by default. At some point there will be a proposal/JEP to flip the switch, something that may create an upgrade challenge for some applications and deployments that haven't embraced the various security features and configuration added over the last 10+ years. The addition of conf/jaxp-strict.properties allows deployments to test more the secure/strict behavior in preparation for a possible future where strict is the default. As examples: trying this out may help identify processing XML that (perhaps unknowingly) makes outbound network connections to fetch DTD, or processing XML that relies on extension functions. So I think very useful to have this configuration available in a shipping JDK but it may need a write-up/JEP before anyone knows about this.

LanceAndersen commented on src/java.xml/share/conf/jaxp-compat.properties:
```diff
@@ -0,0 +1,116 @@
+# prior to a strict configuration such as jaxp-strict.properties. In this
+# configuration, properties that have more restrictive settings as in the
+# strict configuration (jaxp-strict.properties) are reversed back to their
+# defaults. In particular:
```
I think the above needs some more word smithing  as we have not articulated what a strict configuration is or how this differs from jaxp.properties

LanceAndersen commented on src/java.xml/share/conf/jaxp-compat.properties:
```diff
@@ -0,0 +1,116 @@
+#   - JAXP Limits are set to their default values
+#
+# This configuration file can be used to reverse back to a working environment
+# prior to any more restrictive configuration that may have been applied.
```
How does this differ from jaxp.properties for JDK 23?  I can understand for when we move to secure by default, we just need to be clear on the purpose of each jaxp properties files

LanceAndersen reviewed (commented):
Hi Joe,

Overall this looks fine though we need to clarify more as to the differences between jaxp-compat.properties vs jaxp.properties  and also make it clearer why anyone would use jaxp.strict.properties

As part of the review, I would suggest that a Release Note is created which will hopefully clarify when to use which file.

Also when we actually change the defaults in a JDK release to be the equivalent of jaxp-strict.properties, that this file can be removed?  If so this should be documented in the Release Note and perhaps a comment in the properties file itself

JoeWang-Java commented:
Thanks Alan, Lance! I added description of the three config files to both files so that readers can find answers reading any one of them. Also added a release note: https://bugs.openjdk.org/browse/JDK-8330605

naotoj commented on src/java.xml/share/conf/jaxp-strict.properties:
```diff
@@ -0,0 +1,133 @@
+################################################################################
+#           JAXP String Configuration File
```
Nit: typo here: String -> Strict

JoeWang-Java replied:
Thanks!  Fixed.

LanceAndersen commented on src/java.xml/share/conf/jaxp-compat.properties:
```diff
@@ -3,9 +3,27 @@
+# property values when XML factories are initiated.
+#
+# jaxp-strict.properties: this file resembles what will become the Secure-By-Default
+# configuration where a strict restriction is the default. This file allows
```
strict restriction needs rewording.

Perhaps something that indicates that this property file provides settings that will be equivalent to that will be the default JAXP settings in a future release to make the use of JAXP more secure out of the box

LanceAndersen commented on src/java.xml/share/conf/jaxp-compat.properties:
```diff
@@ -3,9 +3,27 @@
+# jaxp-compat.properties: this file can be used to regain compatibility once the
+# JDK has switched to a strict configuration as indicated in jaxp-strict.properties.
+# This configuration contains the same properties as those in jaxp-strict.properties
+# except it sets them back to the current status of the JDK. Note that, although
```
'....sets them back to the current status of the JDK'

I think you are trying to indicate that this property file specifies the JAXP property values that were in place prior to being More Secure?

JoeWang-Java replied:
Thanks Lance for the offline discussion. Updated the description.

seanjmullan commented:
It might just be me, but the word "sample" is too related to programming examples that it makes this feel like something that users may not take seriously. But I think this is something that you really want users to take seriously and try out. Perhaps just drop the word "sample"?

seanjmullan commented:
A few other comments/questions:

Does this need a CSR since you are adding new property files?

Are there any tests to ensure the property files are working correctly?

Also, how does one try out these property files? Is there a system property that needs to be set? Can you add more details to the RN on that?

LanceAndersen commented:
> A few other comments/questions:
> 
> Does this need a CSR since you are adding new property files?

Not sure it does, but Joe will follow up with Joe Darcy
> 
> Are there any tests to ensure the property files are working correctly?

There are tests that Joe added as part of the JDK 22 work for custom config files
> 
> Also, how does one try out these property files? Is there a system property that needs to be set? Can you add more details to the RN on that?

java -Djava.xml.config.file=$JAVA_HOME/conf/jaxp-compat.properties

The property was added in JDK 22 see: https://docs.oracle.com/en/java/javase/22/docs/api/java.xml/module-summary.html#Conf_CF_SP

seanjmullan commented:
> > Also, how does one try out these property files? Is there a system property that needs to be set? Can you add more details to the RN on that?
> 
> java -Djava.xml.config.file=$JAVA_HOME/conf/jaxp-compat.properties
> 
> The property was added in JDK 22 see: https://docs.oracle.com/en/java/javase/22/docs/api/java.xml/module-summary.html#Conf_CF_SP

Sounds good, can you add an example to the RN using the above system property?

JoeWang-Java commented:
> 
> Sounds good, can you add an example to the RN using the above system property?

Added. Thanks.

seanjmullan commented on src/java.xml/share/classes/module-info.java:
```diff
@@ -416,6 +424,37 @@
+ *     <li><a href="#EELimit">JAXP Limits</a> are set to smaller numbers </li>
+ *     </ul>
+ *
+ * This file allows deployments to test the more secure/strict behavior,
+ * identify issues such as a processor unknowingly makes outbound network
+ * connections to fetch DTD, or processes XML that relies on extension functions.
```
Some wording suggestions: 

 ```
* This file allows deployments to test the more secure/strict behavior and
 * identify issues such as a processor unknowingly making outbound network
 * connections to fetch a DTD, or processing XML that relies on extension functions.
```

seanjmullan commented on src/java.xml/share/classes/module-info.java:
```diff
@@ -416,6 +424,37 @@
+ * the property values that are the same as the properties' default values. It can
+ * be used to regain compatibility from a more strict configuration in a future release.
+ * The difference from the default {@code jaxp.properties} is that it contains
+ * additional properties that were not included in {@code jaxp.properties},
```
s/were/are/

AlanBateman commented:
Adding jaxp-strict.properties make sense as it allows developers to identify issues that will arise in the future when XML processing is secure by default. If they deploy with -Djava.xml.config.file=jaxp-strict.properties, and jaxp-strict.properties is removed as part of moving to secure by default, then it should be okay too as the defaults will be strict.

I'm less sure about including jaxp-compat.properties in JDK 23. That's the config file to get temporary relief while you work through the issues with existing code or deployments that break when XML processing is secure by default. Adding in the JDK 23 sends the message that you can "prepare" your command line in advance, which I don't think should be a goal here.

LanceAndersen commented on src/java.xml/share/classes/module-info.java:
```diff
@@ -402,6 +402,14 @@
+ *      <ul>
+ *      <li><a href="#IN_CF_DEFAULT">{@code jaxp.properties}</a></li>
+ *      <li><a href="#IN_CF_STRICT">{@code jaxp-strict.properties}</a></li>
+ *      <li><a href="#IN_CF_COMPAT">{@code jaxp-compat.properties}</a></li>
```
This line should be removed

AlanBateman commented:
/label add build

AlanBateman commented on make/modules/java.xml/Copy.gmk:
```diff
@@ -24,14 +24,18 @@
-XML_LIB_SRC := $(TOPDIR)/src/java.xml/share/conf
-
-$(CONF_DST_DIR)/jaxp.properties: $(XML_LIB_SRC)/jaxp.properties
+$(CONF_DST_DIR)/%: $(JAXPPROPFILE_SRC_DIR)/%
```
The make file changes to copy the properties files look okay but I'm curious about why the naming changes from "XML" to "JAXPPROFILE".

erikj79 replied:
If we are changing this file, we should modernize it.

```
$(eval $(call SetupCopyFiles, COPY_XML_MODULE_CONF, \
    DEST := $(CONF_DST_DIR), \
    FILES := $(wildcard $(TOPDIR)/src/java.xml/share/conf/jaxp*.properties*), \
))

TARGETS += $(COPY_XML_MODULE_CONF)
```

AlanBateman commented on src/java.xml/share/conf/jaxp-strict.properties:
```diff
@@ -0,0 +1,129 @@
+# make the use of JAXP more secure by default. This file allows deployments to
+# test the more secure/strict behavior, identify issues such as a processor
+# unknowingly makes outbound network connections to fetch DTD, or processes XML
+# that relies on extension functions.
```
There isn't a JEP to propose that XML processing be secure by default on the technical roadmap right now so I think this paragraph will need to be tweaked to avoid making any assumptions. I think just say that the file provides the settings for more secure XML processing and drop the text about testing (and "and create your own configuration file for the experiment" from the paragraph below).

AlanBateman commented on src/java.xml/share/classes/module-info.java:
```diff
@@ -416,6 +423,30 @@
+ *     <li><a href="#EELimit">JAXP Limits</a> are set to smaller numbers </li>
+ *     </ul>
+ *
+ * This file allows deployments to test the more secure/strict behavior,
```
I think it might be better to reduce this paragraph down to just say something like "Deploying with this configuation prevents processors from unknowingly making outbound network connections to fetch DTDs, or process XML that makes use of extension functions."

We could say that a future JDK release may use a strict configuration by default but that opens the door to questions as to whether the system property is needed, whether jaxp.propeteries is going away, so maybe better to leave that out for now.

erikj79 commented on make/modules/java.xml/Copy.gmk:
```diff
@@ -24,14 +24,18 @@
 
 ################################################################################
+#
+# Copy property files from share/conf to CONF_DST_DIR LIB_DST_DIR
```
There is no copying to LIB_DST_DIR, so no need to mention it.

JoeWang-Java commented:
Thanks Alan, Erik! Updated accordingly.

erikj79 reviewed (approved):
Build change looks good.

AlanBateman commented on src/java.xml/share/conf/jaxp-strict.properties:
```diff
@@ -0,0 +1,120 @@
+#
+#     java -Djava.xml.config.file=$JAVA_HOME/conf/jaxp-strict.properties
+#
+# The pathname to the configuration file must be valid. If it is not absolute,
```
I think it would be better to drop this paragraph or else just replace it with a sentence to say that the java.xml module description specifies the system property.

JoeWang-Java replied:
Thanks Alan. Replaced the paragraph with a reference to the module description.

AlanBateman reviewed (approved):
Looks good, just one comment on on the jaxp-strict.properties file text.

seanjmullan commented on src/java.xml/share/classes/module-info.java:
```diff
@@ -416,6 +423,30 @@
+ *     </ul>
+ *
+ * Deploying with this configuration prevents processors from unknowingly making
+ * outbound network connections to fetch DTDs, or process XML that makes use of
```
s/process/processing/

AlanBateman replied:
In XML parlance, a "processor" is an aggregation of parsers, serializers, and other things that contribute to the processing. So I think it could be either here, but you have a point and if it stays as "processor" then it should link #FacPro where the term is defined.

seanjmullan replied:
It's the wording that doesn't sound right, before this you have "making" which doesn't sound right with "process". Maybe it needs two sentences.

JoeWang-Java replied:
Thanks! Same as Alan, I thought you were talking about "processor" as well on the first glance, then realized you're referring to the parallel statements :-)

magicus commented on make/modules/java.xml/Copy.gmk:
```diff
@@ -24,14 +24,16 @@
+#
+$(eval $(call SetupCopyFiles, COPY_XML_MODULE_CONF, \
+    DEST := $(CONF_DST_DIR), \
+    FILES := $(wildcard $(TOPDIR)/src/java.xml/share/conf/jaxp*.properties*), \
```
I don't think you need, nor should have, the asterisk after the extension. You are only copying `.properties` files.

```suggestion
    FILES := $(wildcard $(TOPDIR)/src/java.xml/share/conf/jaxp*.properties), \
```

JoeWang-Java replied:
Thanks Magnus!

JoeWang-Java commented:
Update: the plan has changed from providing a configuration file to a template. Module description, test and make file are updated accordingly. Thanks Erik for the new make file!

magicus commented:
I would recommend naming the new file `jaxp-strict.properties.template` instead. This would follow the pattern we have used in the JDK, and I think it is much better at providing clarify as to what this file actually is  -- a template for a `.properties` file.

AlanBateman commented:
Magus's suggestion for the suffix to be .properties.template makes sense, consistent with the one .template that the JDK currently includes in the run-time image.

Overall this looks okay, I'm happy that the implNote update is removed from the proposal as it read too much like the introducing a new supported interface and would have been confusing to have two configurations in the conf directory.

JoeWang-Java commented:
Thanks Magnus, Alan. Renamed the template to jaxp-strict.properties.template.

magicus reviewed (approved):
LGTM

naotoj reviewed (approved):
LGTM

JoeWang-Java commented:
Thanks Sean, Naoto!  Updated accordingly.

JoeWang-Java commented:
/integrate
